crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,818 @@
+///
+/// Examples where these are useful:
+/// - Sorting `__all__` in the global scope,
+/// - Sorting `__slots__` or `__match_args__` in a class scope
+use std::borrow::Cow;
+use std::cmp::Ordering;
+
+use ruff_python_ast as ast;
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::{lexer, Mode, Tok};
+use ruff_python_trivia::leading_indentation;
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use is_macro;
+use itertools::{izip, Itertools};
+
+/// An enumeration of the various kinds of sequences for which Python has
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries).
+///
+/// (I'm aware a set isn't actually a "sequence",
+/// *but* for our purposes it's conceptually a sequence,
+/// since in terms of the AST structure it's almost identical
+/// to tuples/lists.)
+///
+/// Whereas lists, dicts and sets are always parenthesized
+/// (e.g. lists always start with `[` and end with `]`),
+/// single-line tuples *can* be unparenthesized.
+/// We keep the original AST node around for the
+/// Tuple variant so that this can be queried later.
+#[derive(Debug)]
+pub(super) enum SequenceKind<'a> {
+    List,
+    Set,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl SequenceKind<'_> {
+    fn surrounding_parens(&self, source: &str) -> (&str, &str) {
+        match self {
+            Self::List => ("[", "]"),
+            Self::Set => ("{", "}"),
+            Self::Tuple(ast_node) => {
+                if ast_node.is_parenthesized(source) {
+                    ("(", ")")
+                } else {
+                    ("", "")
+                }
+            }
+        }
+    }
+
+    fn opening_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Lsqb,
+            Self::Set => Tok::Lbrace,
+            Self::Tuple(_) => Tok::Lpar,
+        }
+    }
+
+    fn closing_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Rsqb,
+            Self::Set => Tok::Rbrace,
+            Self::Tuple(_) => Tok::Rpar,
+        }
+    }
+}
+
+/// An enumeration of the various kinds of
+/// [display literals](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries)
+/// Python provides for builtin containers.
+#[derive(Debug, is_macro::Is)]
+pub(super) enum DisplayKind<'a> {
+    Sequence(SequenceKind<'a>),
+    Dict { values: &'a Vec<ast::Expr> },
+}
+
+/// Create a string representing a fixed-up single-line
+/// definition of `__all__` or `__slots__` (etc.),
+/// that can be inserted into the
+/// source code as a `range_replacement` autofix.
+pub(super) fn sort_single_line_elements_sequence<F>(
+    kind: &SequenceKind,
+    elts: &[ast::Expr],
+    elements: &[&str],
+    locator: &Locator,
+    mut cmp_fn: F,
+) -> String
+where
+    F: FnMut(&str, &str) -> Ordering,
+{
+    assert_eq!(elts.len(), elements.len());
```

Comment:
Nit: It would be nice if we could encode this constraint by a type that gets returned by the function extracting the `elements`.

Having both `elts` and `elements` wrapped by a new type could also make this function simpler because you could implement it as a method instead (where you only need to provide the locator and comparator)

Reply:
I think he's suggesting creating a dedicated struct like that. Since you're allocating below anyway with `collect_vec`, perhaps you want to do something like:
```rust
struct Whatever<'a>(Vec<(&'a Expr, &'a str)>)
```

That is: store the zipped elements, so that the constraint is always enforced once you've created the struct.

Reply:
Nice -- I gave this a go in https://github.com/astral-sh/ruff/pull/9564/commits/6645934e0f2d3e53071edc1b30a4d17e835ed76e.

I experimented with doing this:

```diff
--- a/crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs
@@ -258,7 +258,7 @@ pub(super) enum SortClassification<'a> {
     /// and it's possible we could generate a fix for it;
     /// here's the values of the elts so we can use them to
     /// generate an autofix:
-    UnsortedAndMaybeFixable { items: Vec<&'a str> },
+    UnsortedAndMaybeFixable { items: SequenceElements<'a> },
     /// The display contains one or more items that are not string
     /// literals.
     NotAListOfStringLiterals,
@@ -309,7 +309,7 @@ impl<'a> SortClassification<'a> {
                 if any_implicit_concatenation {
                     return Self::UnsortedButUnfixable;
                 }
-                return Self::UnsortedAndMaybeFixable { items };
+                return Self::UnsortedAndMaybeFixable { items: SequenceElements::new(items, elements) };
             }
             current = next;
```

Unfortunately it doesn't really work, however, due to the fact that for single-line dicts, we also want to zip the dict-value nodes up along with the dict-key nodes and the string values of the dict keys